KAFKA-8507 kip 499 Unify connection name flag for command line tool

mitchell-h opened the pull request:
This is a change updates ConsoleProducer, ConsumerPerformance, VerifiableProducer, and VerifiableConsumer classes to add and prefer the --bootstrap-server flag for defining the connection point of the Kafka cluster.


### Committer Checklist (excluded from commit message)
- [x ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

mitchell-h committed:
update example reassign file to include log_dirs configuration in alignment with KIP-113

mitchell-h committed:
KIP-499

rondagostino commented on core/src/main/scala/kafka/tools/ConsoleProducer.scala:
```diff
@@ -125,7 +129,11 @@ object ConsoleProducer {
+      .withRequiredArg
+      .describedAs("server to connect to")
+      .ofType(classOf[String])
+    val brokerListOpt = parser.accepts("broker-list", "The broker list string in the form HOST1:PORT1,HOST2:PORT2.")
```
Add "DEPRECATED, use --bootstrap-server instead; ignored if --bootstrap-server is specified"

rondagostino commented on core/src/main/scala/kafka/tools/ConsoleProducer.scala:
```diff
@@ -125,7 +129,11 @@ object ConsoleProducer {
       .ofType(classOf[String])
-    val brokerListOpt = parser.accepts("broker-list", "REQUIRED: The broker list string in the form HOST1:PORT1,HOST2:PORT2.")
+    val bootstrapServerOpt = parser.accepts("bootstrap-server", "REQUIRED: The server(s) to connect to.")
+      .withRequiredArg
```
You can define `bootstrapServerOpt` with `.requiredUnless("broker-list")` as long as you define broker-list first; then the parser will check everything for you automatically.  Also technically not REQUIRED since `--broker-list` is only being deprecated and not removed.  So probably "REQUIRED unless --broker-list is specified" is more accurate.  Also, need to update `ConsoleProducerTest` to test valid and invalid option combinations.

mitchell-h replied:
Updated with suggestion

rondagostino commented on core/src/main/scala/kafka/tools/ConsoleProducer.scala:
```diff
@@ -220,11 +228,28 @@ object ConsoleProducer {
 
     CommandLineUtils.printHelpAndExitIfNeeded(this, "This tool helps to read data from standard input and publish it to Kafka.")
-    CommandLineUtils.checkRequiredArgs(parser, options, topicOpt, brokerListOpt)
+
```
No need to manually check these if you use `.requiredUnless("broker-list")`.

mitchell-h replied:
Updated with requiredUnless

rondagostino commented on core/src/main/scala/kafka/tools/ConsoleProducer.scala:
```diff
@@ -220,11 +228,28 @@ object ConsoleProducer {
 
     val topic = options.valueOf(topicOpt)
+
+    val bootstrapServer = options.valueOf(bootstrapServerOpt)
```
Probably want to do something like this:

```
val brokerHostsAndPorts = options.valueOf(if (options.has(bootstrapServerOpt)) bootstrapServerOpt else brokerListOpt)
ToolsUtils.validatePortOrDie(parser, brokerHostsAndPorts)
```

mitchell-h replied:
Updated with suggestion

rondagostino reviewed (commented):
Reviewed the first file.  Generally need to add tests all around.

hachikuji commented on tools/src/main/java/org/apache/kafka/tools/VerifiableConsumer.java:
```diff
@@ -503,10 +503,17 @@ private static ArgumentParser argParser() {
                 .newArgumentParser("verifiable-consumer")
                 .defaultHelp(true)
                 .description("This tool consumes messages from a specific topic and emits consumer events (e.g. group rebalances, received messages, and offsets committed) as JSON objects to STDOUT.");
+        parser.addArgument("--bootstrap-server")
```
We also want to deprecate the old options. Typically we just prefix the description with `DEPRECATED`. See for example, the `--zookeeper` argument use in `TopicCommand`. We could also print a warning, but it may be overkill.

mitchell-h replied:
Updated

hachikuji commented:
One more thing. I grepped `docs/` for uses of these old args and found a few hits. Can we update the docs as well?

mitchell-h committed:
KIP-499 - Updated inline with commit comments.  Added requiredUnless options.  simplified validateOrDie logic.

chia7712 commented on core/src/main/scala/kafka/tools/ConsoleProducer.scala:
```diff
@@ -125,10 +129,15 @@ object ConsoleProducer {
       .withRequiredArg
       .describedAs("broker-list")
       .ofType(classOf[String])
+    val bootstrapServerOpt = parser.accepts("bootstrap-server", "REQUIRED unless --broker-list(deprecated) is specified. The server(s) to connect to.")
```
The document "The broker list string in the form HOST1:PORT1,HOST2:PORT2." is useful to terminal user. Could we keep it?

mitchell-h replied:
Fixed and updated.  Thanks.

mitchell-h committed:
Updated /docs to use `kafka-console-producer --bootstrap-server` instead of --broker-list

added 'The broker list string in the form HOST1:PORT1,HOST2:PORT2' to command line usage for kafka-console-producer

stanislavkozlovski commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -202,9 +203,14 @@ object ConsumerPerformance extends LazyLogging {
+      .withRequiredArg
+      .describedAs("broker-list")
+      .ofType(classOf[String])
+    val bootstrapServerOpt = parser.accepts("bootstrap-server", "REQUIRED unless --broker-list(deprecated) is specified. The server(s) to connect to.")
```
Should we also add `The broker list string in the form HOST1:PORT1,HOST2:PORT2.` to this message?

stanislavkozlovski commented on core/src/test/scala/unit/kafka/tools/ConsoleProducerTest.scala:
```diff
@@ -37,20 +37,45 @@ class ConsoleProducerTest {
     "key.separator=#"
   )
-
+  val bootStrapServerValidArgs: Array[String] = Array(
```
nit: camelCase naming would be `bootstrapServerValidArgs`

stanislavkozlovski commented on core/src/test/scala/unit/kafka/tools/ConsoleProducerTest.scala:
```diff
@@ -37,20 +37,45 @@ class ConsoleProducerTest {
     "--t", // not a valid argument
     "t3"
   )
+  val bootStrapserverOverride: Array[String] = Array(
```
nit: ditto on the camel case

stanislavkozlovski commented:
Thanks a lot for the work on this! I've got a couple of comments:


1. `vagrant/README.md` mentions
```
bin/kafka-console-producer.sh --broker-list broker1:9092,broker2:9092,broker3:9092 --topic sandbox
```

2. Multiple system tests leveraging the commands mention broker-list. Should we change those?

3. There are more tools to update from --broker-list to --bootstrap-server. Did we miss them in the KIP and if so, is it worth adding them in? (kafka-reassign-partitions.sh, GetOffsetShell, ReplicaVerificationTool, TransactionalMessageCopier, VerifiableLog4jAppender)

hachikuji commented:
ok to test

hachikuji committed:
Fix checkstyle

hachikuji committed:
Fix test cases in ConsoleProducerTest

hachikuji commented:
retest this please

hachikuji commented:
https://builds.apache.org/job/kafka-pr-jdk11-scala2.13/4697/

hachikuji reviewed (approved):
LGTM. Note I pushed a couple trivial tweaks to get this over the line.

hachikuji commented:
We will follow up in a separate PR with @stanislavkozlovski 's suggestions.
